smalld/smalld.py

```diff
@@ -108,25 +118,55 @@ def run(self):
 
             self.gateway = Gateway(gateway_url)
 
-            for data in self.gateway:
-                logger.debug("gateway payload received: %s", data)
-                for listener in self.listeners:
-                    listener(AttrDict(json.loads(data)))
+            try:
+                for data in self.gateway:
+                    logger.debug("gateway payload received: %s", data)
+                    for listener in self.listeners:
+                        listener(data)
+            except GatewayClosedException as e:
+                if e.code in recoverable_error_codes:
+                    logger.error(f"gateway closed: {e}")
+                    self.reconnect()
+                else:
+                    raise
 
             time.sleep(5)
 
 
+class GatewayClosedException(Exception):
+    def __init__(self, code, reason):
+        super().__init__(f"{code}: {reason}")
+        self.code = code
+        self.reason = reason
+
+
 class Gateway:
+    FRAME_OPCODE_TEXT = 0x1
+    FRAME_OPCODE_CLOSE = 0x8
```

Comment:
Why do we define these rather than use the constants from the websocket-client lib?